Autoscaling Warm Pool Support for K8S Variants

parth3sh opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**
#1788 


**Description of changes:**
**Goal: To support EC2 Autoscaling “warm pools” in K8S Bottlerocket variants.**

- Add autoscaling.should-wait setting to K8S variants. Handle migrations for setting in case of downgrade.
- Create "hachiko", a helper program that retrieves the autoscaling.should-wait setting and autoscaling lifecycle state of host. If this setting is true, program waits until lifecycle state is "InService".
- autoscaling.should-wait setting is written to a config file (according to hachiko-toml template) during instance boot and is later parsed by the helper program.
- Create hachiko.service to execute helper program prior to the running of Kubelet in order to prevent warm pool instances from running Kubelet and falling into the "NotReady" state.

**Testing done:**

- Program is installed properly.
- Service runs as expected and at the right point during host configuration.
- Waiting program tested locally and on a launched bottlerocket instance for ability to correctly parse config file and retrieve lifecycle state of the host.
- Config file with autoscaling.should-wait setting value is correctly generated and parsed. 

**Testing required:**
- Confirm autoscaling.should-wait is migrated properly by confirming it is added to k8s variants upon upgrade to version 10.0.0 and removed upon downgrade.

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

cbgbt commented on sources/Cargo.toml:
```diff
@@ -68,6 +68,7 @@ members = [
     "api/migration/migrations/v1.9.0/kernel-modules-setting",
     "api/migration/migrations/v1.9.0/kernel-modules-setting-metadata",
     "api/migration/migrations/v1.9.0/kubelet-no-daemon-reload",
+    "api/migration/migrations/v1.10.0/add-eks-autoscaling",
     "api/migration/migrations/v1.9.0/updates-targets-base-url",
```
This migration should be ordered below the ones for `1.9.0`.

It also seems like this may be included in the wrong commit.

parth3sh replied:
I have moved that migration below the 1.9.0 migrations and shifted the change to ["Handle migrations for eks-autoscaling.should-wait"](https://github.com/bottlerocket-os/bottlerocket/pull/2322/commits/c173ed58cc568c2664e44db6fd36520b347aa5be)

cbgbt commented on sources/ascwp/Cargo.toml:
```diff
@@ -0,0 +1,24 @@
+toml = "0.5.1"
+tokio = { version = "~1.14", default-features = false, features = ["macros", "rt-multi-thread"] }
+rusoto_core = { version = "0.48.0", default-features = false, features = ["rustls"] }
+rusoto_cloudformation = { version = "0.48.0", default-features = false, features = ["rustls"] }
```
These dependencies can likely be removed.

cbgbt commented on sources/ascwp/Cargo.toml:
```diff
@@ -0,0 +1,24 @@
+async-std = "1.12.0"
+
+[dev-dependencies]
+httptest = "0.15"
```
Seems like this dependency can also be removed.

cbgbt commented on sources/ascwp/Cargo.toml:
```diff
@@ -0,0 +1,24 @@
+rusoto_cloudformation = { version = "0.48.0", default-features = false, features = ["rustls"] }
+imdsclient = { path = "../imdsclient", version = "0.1.0" }
+hyper = "0.14.2"
+async-std = "1.12.0"
```
These also seem to not be used.

cbgbt commented on sources/ascwp/src/config.rs:
```diff
@@ -0,0 +1,60 @@
+
+impl Config {
+    //uses default config path
+    pub(crate) fn new() -> Result<Self> {
```
For functions, can you use docstring comments? They begin with `/// `. These comments generally use capitalized sentences with punctuation. (Same should be true for other functions in the PR)

Commit:
Add waiting helper program to run before Kubelet

Add hachiko program which parses config file with
eks-autoscaling.should-wait setting of host and if setting is true waits
for autoscaling lifecycle state to be "InService". Add
fetch_lifecycle_state() function to imdsclient to retrieve autoscaling
lifecycle state of host.

Commit:
Add autoscaling.should-wait setting to k8s variants

Add autoscaling.should-wait setting to k8s variants. Set default
value to false. Define path for config file and template for setting
value parsed by hachiko helper program.

Commit:
Handle migrations for autoscaling.should-wait

Handle adding autoscaling.should-wait setting in v1.10.0 and removing setting in case of
bottlerocket version downgrade.

cbgbt commented on sources/ascwp/src/config.rs:
```diff
@@ -0,0 +1,60 @@
+    //uses path passed into function
+    pub(crate) fn from_file<P: AsRef<Path>>(path: P) -> Result<Self> {
+        let path = path.as_ref();
+        let s = fs::read_to_string(path).context(error::ConfigReadSnafu { path })?;
```
I would avoid single character variable names.

cbgbt commented on sources/ascwp/src/main.rs:
```diff
@@ -0,0 +1,59 @@
+//Path to config file containing host's eks_autoscaling.should-wait setting
+const CONFIG_PATH: &str = "/etc/ascwp.toml";
+//Marker file that is created after first run. We only want this program to run once.
+const MARKER_FILE: &str = "/var/lib/bottlerocket/ascwp.ran";
```
I think it would be good to make this configurable via `ascwp.toml`

cbgbt commented on sources/ascwp/src/main.rs:
```diff
@@ -0,0 +1,59 @@
+
+
+
+//Path to config file containing host's eks_autoscaling.should-wait setting
```
For non docstring comments as well, I'd add a space between `//` and the start of the comment.

cbgbt commented on sources/ascwp/src/main.rs:
```diff
@@ -0,0 +1,59 @@
+        })
+}
+
+//Continously fetches lifecycle state of the host. Returns when host is "InService"
```
nit: `Continuously`

cbgbt commented on sources/ascwp/src/main.rs:
```diff
@@ -0,0 +1,59 @@
+    let mut lifecycle_state = get_lifecycle_state(&mut client).await?;
+    while lifecycle_state.ne("InService") {
+        lifecycle_state = get_lifecycle_state(&mut client).await?;
+        task::sleep(Duration::from_secs(5)).await;
```
I believe we typically use `tokio`'s [sleep implementation](https://docs.rs/tokio/latest/tokio/time/fn.sleep.html). You may be able to remove an additional dependency that way.

Separately: The poll interval should be hoisted to a `CONSTANT`, or perhaps even moved to the config file (with a default value).

cbgbt commented on sources/ascwp/src/config.rs:
```diff
@@ -0,0 +1,60 @@
+use snafu::ResultExt;
+use std::fs;
+use std::path::{Path, PathBuf};
+const DEFAULT_CONFIG_PATH: &str = "/testConfig.toml";
```
It looks like we set a separate default below. Should we remove this one? Or centralize the default to this file?

cbgbt commented on sources/imdsclient/src/lib.rs:
```diff
@@ -1,19 +1,14 @@
 /*!
```
It looks like some whitespace may have been accidentally removed in this docstring comment.

cbgbt commented on sources/models/src/aws-k8s-1.22/mod.rs:
```diff
@@ -27,4 +27,5 @@ struct Settings {
     container_registry: RegistrySettings,
     oci_hooks: OciHooks,
     cloudformation: CloudFormationSettings,
+    eks_autoscaling: AutoScalingSettings,
```
I would just call this setting `autoscaling` so that it is consistent with the ECS setting.

cbgbt commented on packages/os/os.spec:
```diff
@@ -423,13 +434,13 @@ install -d %{buildroot}%{_cross_datadir}/updog
 
 install -d %{buildroot}%{_cross_templatedir}
-install -p -m 0644 %{S:5} %{S:6} %{S:7} %{S:8} %{buildroot}%{_cross_templatedir}
+install -p -m 0644 %{S:5} %{S:6} %{S:7} %{S:8} %{S:9} %{S:10} %{buildroot}%{_cross_templatedir}
```
It looks like we're adding a reference here to sources 9 and 10, but only 10 is new in this PR. Is that right?

cbgbt commented on sources/api/migration/migrations/v1.10.0/add-eks-autoscaling/src/main.rs:
```diff
@@ -0,0 +1,21 @@
+/// Remove the whole `settings.eks_autoscaling` prefix if we downgrade.
+fn run() -> Result<()> {
+    migrate(AddPrefixesMigration(vec!["settings.autoscaling"]))
+}
```
We'll want this migration to only run on EKS variants. You *should* be able to do something similar to [this](https://github.com/bottlerocket-os/bottlerocket/blob/2a6c5e92352134bb2c1e17c5317d1446b97ad7f5/sources/api/migration/migrations/v1.8.0/pki-affected-services/src/main.rs#L18). You'll want to test the migration on k8s and ECS variants, though.

cbgbt commented on sources/api/migration/migrations/v1.10.0/add-eks-autoscaling/src/main.rs:
```diff
@@ -0,0 +1,21 @@
+use std::process;
+
+/// We added a new setting prefix for configuring autoscaling in k8s variants.
+/// Remove the whole `settings.eks_autoscaling` prefix if we downgrade.
```
nit:

```suggestion
/// Remove the whole `settings.autoscaling` prefix if we downgrade.
```

cbgbt commented on sources/hachiko/src/config.rs:
```diff
@@ -0,0 +1,57 @@
+        let config = Config::from_file(&path).unwrap();
+        assert!(config.should_wait);
+    }
+}
```
The whitespacing on this file is unusual in a few places, and it's missing a [newline at the end of the file](https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline).

I think running `cargo fmt` in the `hachiko` directory should fix this.

cbgbt commented on sources/hachiko/src/main.rs:
```diff
@@ -0,0 +1,66 @@
+use simplelog::{Config as LogConfig, SimpleLogger};
+use tokio::time::{sleep, Duration};
+
+/// Path to config file containing host's autoscaling.should-wait setting
```
I'm actually not sure if docstrings work for constants. You can check by running `cargo doc` and peeking at the output. I tend to just use regular comments.

```suggestion
// Path to config file containing host's autoscaling.should-wait setting
```

gthao313 commented on sources/hachiko/src/main.rs:
```diff
@@ -56,11 +56,22 @@ async fn get_config() -> Result<Config> {
     SimpleLogger::init(LevelFilter::Info, LogConfig::default())
     .expect("unable to configure logger");
-    let should_wait_value = get_config().await.unwrap().should_wait;
+    let config_parse = get_config().await.unwrap();
```
`unwrap` leaves us with a less meaningful output than explicit handling, so it would be better to use `context`  which is intended to be a human-readable, user-facing explanation for the error that has occurred.

gthao313 replied:
`unwrap` leaves us with a less meaningful output than explicit handling, so it would be better to use `context`  which is intended to be a human-readable, user-facing explanation for the error that has occurred.

gthao313 commented on sources/hachiko/src/main.rs:
```diff
@@ -56,11 +56,22 @@ async fn get_config() -> Result<Config> {
         "autoscaling.should-wait value is {}", should_wait_value
     );
     if should_wait_value{
         wait_until_inservice().await.unwrap();
```
Same as above `unwrap`

Commit:
Add service to run hachiko program before Kubelet

Configure hachiko.service to run hachiko after settings applier and
network-online and before Kubelet. Create template for config file containing autoscaling.should-wait
setting. Automate hachiko package buildig in os.spec.

zmrow commented on packages/os/hachiko.service:
```diff
@@ -0,0 +1,15 @@
+# after a successful run.
+ConditionPathExists=!/var/lib/bottlerocket/hachiko.ran
+
+[Service]
```
I think we probably want `RemainAfterExit=true` for this oneshot service so systemd doesn't show it as "deactivating" or "dead" after it runs.

See the service unit for `early-boot-config` for example.

zmrow commented on packages/os/os.spec:
```diff
@@ -546,6 +558,7 @@ install -p -m 0644 %{S:300} %{buildroot}%{_cross_udevrulesdir}/80-ephemeral-stor
 %{_cross_unitdir}/metricdog.timer
 %{_cross_unitdir}/send-boot-success.service
 
+
```
Is this a spurious extra space?

zmrow reviewed (commented):
Just a few nits.

It sounds like you're already testing the migration; once that's done please add it to the PR description!

zmrow commented on sources/hachiko/src/main.rs:
```diff
@@ -0,0 +1,71 @@
+#[tokio::main]
+async fn main() {
+    SimpleLogger::init(LevelFilter::Info, LogConfig::default())
+    .expect("unable to configure logger");
```
Let's use the `.context(snafu_error_here)` to provide context rather than using `expect()`

gthao313 commented on sources/hachiko/src/config.rs:
```diff
@@ -0,0 +1,58 @@
+    "#;
+
+    // Example config file where user sets autoscaling.should-wait to true.
+    const true_should_wait: &str = r#"
```
`Const` variable needs upper case
```suggestion
    const TRUE_SHOULD_WAIT: &str = r#"
```

arnaldo2792 commented on sources/hachiko/src/config.rs:
```diff
@@ -0,0 +1,55 @@
+    use tempfile::TempDir;
+
+    // Example config file where user sets autoscaling.should-wait to false.
+    const false_should_wait: &str = r#"
```
Same as below, the constant name should be all `UPPER_CASE`

arnaldo2792 commented on packages/os/hachiko.service:
```diff
@@ -0,0 +1,16 @@
+[Unit]
+Description= Waits for host to be in-service before running kubelet if autoscaling.should_wait is true.
```
Descriptions in systemd units are "especial". This is what systemd will print to the journal when it attempts to start a unit, so a simpler, shorter, message is preferred. Look at other systemd units for inspiration.

arnaldo2792 commented on packages/os/hachiko.service:
```diff
@@ -0,0 +1,16 @@
+RemainAfterExit=true
+
+[Install]
+WantedBy=kubelet.service
```
I looked at all the other services in Bottlerocket and none of them use `*.service` here. Thus, I think a target should be used, and for this case `multi-user.target` should be adequate.

arnaldo2792 commented on packages/os/os.spec:
```diff
@@ -232,6 +236,11 @@ Summary: Settings generator for ECS
 %description -n %{_cross_os}pluto
 %{summary}.
+
+%package -n %{_cross_os}hachiko
+Summary: Autoscaling warmpool support
+%description -n %{_cross_os}hachiko
+%{summary}.
 %endif
```
The conditionals to add packages depending on the variant being built, change, you need to rebase `develop` to bring those changes.

arnaldo2792 commented on packages/os/os.spec:
```diff
@@ -423,13 +434,13 @@ install -d %{buildroot}%{_cross_datadir}/updog
   %{S:100} %{S:101} %{S:102} %{S:103} %{S:105} \
   %{S:106} %{S:107} %{S:110} %{S:111} %{S:112} \
-  %{S:113} %{S:114} %{S:118} %{S:119} \
+  %{S:113} %{S:114} %{S:118} %{S:119} %{S:120} \
```
The `hachiko.service` should only be installed for `aws` variants, this is why the build is failing for VMware and Metal variants, since this service in packages conditionally a few lines below.

arnaldo2792 commented on packages/selinux-policy/fs.cil:
```diff
@@ -47,6 +47,7 @@
 (filecon "/.*/usr/bin/storewolf" file api_exec)
 (filecon "/.*/usr/bin/cfsignal" file api_exec)
 (filecon "/.*/usr/bin/thar-be-settings" file api_exec)
+(filecon "/.*/usr/bin/hachiko" file api_exec)
```
@bcressey there are a few programs now that are `api_exec` just to let them write to the persistent partition. Do you think it will be worth having a less privileged label that grants this permission for programs like `cfsignal` or `hachiko`?

arnaldo2792 commented on sources/hachiko/src/main.rs:
```diff
@@ -0,0 +1,58 @@
+mod config;
```
This program is modified in two commits, I think it will be best to squash those two commits since in commit 6d3efecc57f83c2b9a72e63bdb8c20bf4f038ee3, the program uses `.expect` to handle errors, whereas in commit a0764b084f31601f7e05eb66f88a076b54d68a1e Snafu contexts are used.

arnaldo2792 commented on sources/hachiko/src/main.rs:
```diff
@@ -0,0 +1,58 @@
+    while lifecycle_state.ne("InService") {
+        lifecycle_state = get_lifecycle_state(&mut client).await?;
+        info!("Lifecycle state is {} ... waiting.", lifecycle_state);
+        sleep(Duration::from_secs(5)).await;
```
Shouldn't this `await` include `?`?

gthao313 replied:
I don't think so : ). standard syntax of `tokio::time::sleep` is  `sleep(Duration).await;`. https://docs.rs/tokio/latest/tokio/time/fn.sleep.html

arnaldo2792 commented on packages/selinux-policy/fs.cil:
```diff
@@ -47,6 +47,7 @@
 (filecon "/.*/usr/bin/storewolf" file api_exec)
```
For the lack of a better place to leave this comment. This commit message isn't properly formatted.

gthao313 commented:
replace by https://github.com/bottlerocket-os/bottlerocket/pull/2562
